channels/generic/websocket.py

```diff
@@ -254,3 +256,155 @@ class AsyncJsonWebsocketConsumer(AsyncWebsocketConsumer):
     @classmethod
     async def encode_json(cls, content):
         return json.dumps(content)
+
+
+class AsyncJsonWebsocketDemultiplexer(AsyncJsonWebsocketConsumer):
+
+    """
+    JSON-understanding WebSocket consumer subclass that handles demultiplexing
+    streams using a "stream" key in a top-level dict and the actual payload
+    in a sub-dict called "payload". This lets you run multiple streams over
+    a single WebSocket connection in a standardised way.
+    Incoming messages on streams are dispatched to consumers so you can
+    just tie in consumers the normal way.
+
+    Set a mapping of streams to an application classes in the "applications"
+    keyword.
+    """
+
+    async def receive_json(self, content, **kwargs):
+        """
+        Rout the message to the correct child.
+        """
+
+        # Check the frame looks good
+        if isinstance(content, dict) and "stream" in content and "payload" in content:
+            # Match it to a channel
+            stream = content["stream"]
+            try:
+                child_queue = await self._get_child_queue(stream)
+
+            except ValueError:
+                raise ValueError(
+                    "Invalid multiplexed frame received (stream not mapped)"
+                )
+
+            payload = content["payload"]
+
+            # send it on to the child application
+            child_queue.put_nowait(
+                {
+                    "type": "websocket.receive",
+                    "text": await self.encode_json(payload)
+                }
+            )
+            return
+
+        else:
+            raise ValueError("Invalid multiplexed **frame received (no channel/payload key)")
+
+    async def websocket_send(self, message, child_lookup):
+        """
+        Capture downstream websocket.send messages from child applications.
+        """
+        text = message.get("text")
+
+        json = await self.decode_json(text)
+        data = {
+            "stream": child_lookup,
+            "payload": json
+        }
+
+        await self.send_json(data)
+
+    ## De-multiplexing methods
+
+    applications = {}  # type: Dict[str: Type[channels.consumer.AsyncConsumer]]
+
+    def __init__(self, scope: Dict[str, Any]):
+        super().__init__(scope)
+        self.child_upstream_queues = {}  # type: Dict[str, asyncio.Queue]
+
+    async def __call__(self, receive, send):
+        """
+        Set up child applications and set up self.
+        """
+        child_tasks = []
+        try:
+            self.child_upstream_queues, child_tasks = await self._init_child_applications()
```

Comment:
There's no need to make `init_child_applications` a separate method here really, it just makes it harder to read.